qualtran/bloqs/mod_arithmetic/_shims.py

```diff
@@ -0,0 +1,250 @@
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from collections import defaultdict
+from functools import cached_property
+from typing import Set
+
+from attrs import frozen
+
+from qualtran import Bloq, QBit, QUInt, Register, Signature, Soquet
+from qualtran.bloqs.arithmetic import Add, AddK
+from qualtran.bloqs.arithmetic._shims import CHalf, Lt, MultiCToffoli, Negate, Sub
+from qualtran.bloqs.basic_gates import CNOT, CSwap, Swap, Toffoli
+from qualtran.drawing import Circle, TextBox, WireSymbol
+from qualtran.resource_counting.symbolic_counting_utils import log2
+
+
+@frozen
+class ModAdd(Bloq):
+    n: int
+    mod: int
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature([Register('x', QUInt(self.n)), Register('y', QUInt(self.n))])
+
+
+@frozen
+class ModSub(Bloq):
+    n: int
+    mod: int
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature([Register('x', QUInt(self.n)), Register('y', QUInt(self.n))])
+
+
+@frozen
+class CModSub(Bloq):
+    n: int
+    mod: int
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature(
+            [Register('ctrl', QBit()), Register('x', QUInt(self.n)), Register('y', QUInt(self.n))]
+        )
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+        # Roetteler
+        return {(Toffoli(), 16 * self.n * log2(self.n) - 23.8 * self.n)}
+
+    def wire_symbol(self, soq: 'Soquet') -> 'WireSymbol':
+        if soq.reg.name == 'ctrl':
+            return Circle()
+        elif soq.reg.name == 'x':
+            return TextBox('x')
+        elif soq.reg.name == 'y':
+            return TextBox('x-y')
+
+    def __str__(self):
+        return self.__class__.__name__
+
+
+@frozen
+class CModAdd(Bloq):
+    n: int
+    mod: int
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature(
+            [Register('ctrl', QBit()), Register('x', QUInt(self.n)), Register('y', QUInt(self.n))]
+        )
+
+
+@frozen
+class _ModInvInner(Bloq):
+    n: int
+    mod: int
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature([Register('x', QUInt(self.n)), Register('out', QUInt(self.n))])
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+        listing = [
+            (MultiCToffoli(self.n + 1), 1),
+            (CNOT(), 1),
+            (Toffoli(), 1),
+            (MultiCToffoli(n=3), 1),
+            (CNOT(), 2),
+            (Lt(self.n), 1),
+            (CSwap(self.n), 2),
+            (Sub(self.n), 1),
+            (Add(QUInt(self.n)), 1),
+            (CNOT(), 1),
+            (ModDbl(self.n, self.mod), 1),
+            (CHalf(self.n), 1),
+            (CSwap(self.n), 2),
+            (CNOT(), 1),
+        ]
+        summer = defaultdict(lambda: 0)
+        for bloq, n in listing:
```

Comment:
This is because you can accidentally have duplicate bloqs in the listing which won't get counted if you just convert directly into a set right? This caught be out before.

Reply:
added a note, but the listing follows the paper figure which indeed could have duplicate "keys"